docs/development.md

```diff
@@ -0,0 +1,153 @@
+
+* Production: `ProductionAPIs`
+* Staging: `StagingAPIs`
+
+![AWS Architecture](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.githubusercontent.com/LBHackney-IT/coronavirus-frontdoor-snapshot/master/docs/infrastructure.iuml)
+
+
+## Integration with other tools
+
+There is currently an integration with the "I Need Help" tool.
+
+While the "I Need Help" tool has been replaced by "Here To Help", the tool remains accessible.
+
+It has an integration with Better Conversations as documented [here](https://docs.google.com/document/d/1jzpirzR3U8pDOwy_H-sI3jtZi9Or_2nPQH6nCW1lEuY/edit#heading=h.u4foorjj4xue)
+
+## Getting started
+This project uses **yarn** for dependency management and is built with Next.js.
+
+1. Install the project dependencies
+   ```
+   yarn
+   ```
+   and
+   `mkdir dynamodblocal`
+   if the 'dynamodblocal' directory does not exist.
+2. Create a `.env` file based off of the `.env.sample` that exists.
+3. [Set up DynamoDB Local](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBLocal.html)
+   or start a local dynamodb container with Docker like so: `docker-compose up`
+4. Configure AWS cli with `aws configure`
+5. Create local DynamoDB plans table
+   ```bash
+   aws dynamodb create-table --cli-input-json file://./config/tables/referrals.json --endpoint-url http://localhost:8000
+   aws dynamodb create-table --cli-input-json file://./config/tables/conversations.json --endpoint-url http://localhost:8000 > /dev/null
+   ```
+6. Start running your local copy of understanding vulnerability.
+   ```
+   yarn dev
+   ```
+
+7. (Optional) add some test data:
+```
+curl -X POST \
+  http://localhost:3000/api/snapshots \
+  -H 'content-type: application/json' \
+  -d '{
+"firstName" : "Sue",
+"lastName" : "Taylor",
+"dob": {},
+"systemIds": ["123"],
+"createdBy": "",
+"postcode": "E1 6AW"
+}'
+```
+which should return a `201` with a snapshot id like: `rrCN4o37`. Then browse to `http://localhost:3000/snapshots/rrCN4o37` to verify the page opens, if that is the case your local setup is complete.
+
+## Working with DynamoDB locally
+If you need to view, edit or delete data from your local copy of DynamoDB when working on a feature
+there is a useful admin tool you can run.
+
+```(bash)
+yarn dynamo-admin
+```
+
+## Testing
+
+### Unit tests
+Unit tests are written with Jest, we are using `@testing-library/react` for React component testing.
+
+```(bash)
+yarn unit-test
+```
+
+### Integration tests
+Integration tests are written with Cypress.
+
+To run integration tests:
+
+1. Set these env variables to the following values:
+```(bash)
+INH_URL=https://inh-admin-test.hackney.gov.uk
+AIRTABLE_API_KEY=xx
+AIRTABLE_PROMPTS_TABLE_NAME="Conversational Prompts"
+FSS_PUBLIC_API_URL=http://localhost:8085/fss
+AIRTABLE_BASE_ID=baseId123
+AIRTABLE_TABLE_NAMES="Service directory"
+AIRTABLE_BASE_URL=http://localhost:8085
+AIRTABLE_PROMPTS_API_KEY=xx
+AIRTABLE_PROMPTS_BASE_ID=baseId456
+HOTJAR_ID=12345
+```
```

Comment:
Is this list of variables up to date? I think I remember having issues running integration tests because some GovNotify environment variable was missing.

The questions is, are there any more that have been added that are necessary?

Reply:
Yep, good question. I'm just checking through this now. 

A fun little script you can run to see all the env vars that are used throughout the code base:

```
 % git grep -h --only-matching --extended-regexp "process.env.([a-zA-Z_-]+)"|sort|uniq|sed 's/process.env.//'                                                                                                hackney(h_dev)
ADDRESSES_API_KEY
ADDRESSES_API_URL
AIRTABLE_API_KEY
AIRTABLE_BASE_ID
AIRTABLE_BASE_URL
AIRTABLE_PROMPTS_API_KEY
AIRTABLE_PROMPTS_BASE_ID
AIRTABLE_PROMPTS_TABLE_NAME
AIRTABLE_TABLE_NAMES
ALLOWED_GROUPS
BASIC_ALLOWED_GROUPS
CONVERSATIONS_TABLE_NAME
ENV
FSS_PUBLIC_API_URL
GOV_NOTIFY_API_KEY
HOTJAR_ID
INH_URL
JWT_SECRET
NEXT_PUBLIC_API_URL
NEXT_PUBLIC_ENV
NEXT_PUBLIC_GTM_ID
NEXT_PUBLIC_URL
PORT
REFERRAL_EMAIL_TEMPLATE_ID
REFERRAL_RECEIPT_EMAIL_TEMPLATE_ID
REFERRAL_RESIDENT_EMAIL_TEMPLATE_ID
REFERRAL_RESIDENT_SMS_TEMPLATE_ID
SHOW_TOPIC_EXPLORER
SIGNPOSTL_EMAIL_TEMPLATE_ID
TABLE_NAME
_HANDLER
```

Or this command to find where they are:
```
 % git grep --only-matching --extended-regexp "process.env.([a-zA-Z_-]+)"                                                                                                                                  hackney(h_dev)
```

Reply:
Just now comparing that with what is defined in the serverless.yml file.. These are taken care of during the compile time from CircleCI.

* `ADDRESSES_API_KEY`
* `ADDRESSES_API_URL`
* `AIRTABLE_BASE_URL`
* `HOTJAR_ID`
* `NEXT_PUBLIC_ENV`
* `NEXT_PUBLIC_GTM_ID`
* `NEXT_PUBLIC_SIGNPOST_ENABLED`

With exception to 
* `PORT` - Default to 3000
* `_HANDLER` - No real value and can be removed from code base

Reply:
Unfortunately it looks like there's a few different ways of initialising the env vars here. Ideally we'd only be providing them at run time, rather than baking them in at compile time.

One other thing to note, from memory the NEXT_PUBLIC_* env vars need to be baked in because they're collapsed by webpack during transpilation. The others should be OK.

Nothing needs to be done on this at the moment.

Reply:
@elena-vi / @kosiakkatrina do you have the env vars that need to be set for running the integration tests? I think there are some missing from this list defined under "Integration tests" in particular, compared to the list defined in the circleci configuration for integration tests:

https://github.com/LBHackney-IT/coronavirus-frontdoor-snapshot/blob/34fc975b4e79afe50cadd21bfaa850c4f3b89e38/.circleci/config.yml#L74-L99

Reply:
Most of the test env vars are in .env.sample
```
INH_URL=https://inh-admin-test.hackney.gov.uk 
AIRTABLE_API_KEY=xx
AIRTABLE_PROMPTS_TABLE_NAME="Conversational Prompts"
FSS_PUBLIC_API_URL=http://localhost:8085/fss
AIRTABLE_BASE_ID=baseId123
AIRTABLE_TABLE_NAMES="Service directory"
AIRTABLE_BASE_URL=http://localhost:8085
AIRTABLE_PROMPTS_API_KEY=xx
AIRTABLE_PROMPTS_BASE_ID=baseId456
HOTJAR_ID=12345
REFERRAL_EMAIL_TEMPLATE_ID=xxx
REFERRAL_RECEIPT_EMAIL_TEMPLATE_ID=xxx
REFERRAL_RESIDENT_EMAIL_TEMPLATE_ID=xyz
SIGNPOSTL_EMAIL_TEMPLATE_ID=123
REFERRAL_RESIDENT_SMS_TEMPLATE_ID=abc
GOV_NOTIFY_API_KEY=xxxx
NEXT_PUBLIC_ENV=test
TABLE_NAME=referrals
CONVERSATIONS_TABLE_NAME=conversations
```

Reply:
Probably this as well (depends on the port you're running on)
```
NEXT_PUBLIC_URL=http://localdev.hackney.gov.uk:3000
NEXT_PUBLIC_API_URL=http://localdev.hackney.gov.uk:3000/api
```

Reply:
We also need this:
```
SHOW_TOPIC_EXPLORER=true
```

Also for the time being this needs to be set, but it should be irrelevant when we remove the feature flag:
```
NEXT_PUBLIC_SIGNPOST_ENABLED=true
```

That's it.

Reply:
Great thanks Kat - have added those final two to the .env.sample. I'll make a note about changing the public urls based on configuration

Reply:
> Probably this as well (depends on the port you're running on)
> 
> ```
> NEXT_PUBLIC_URL=http://localdev.hackney.gov.uk:3000
> NEXT_PUBLIC_API_URL=http://localdev.hackney.gov.uk:3000/api
> ```

Good call! On the topic of these env vars,... do we have it documented anywhere of how to make `localdev.hackney.gov.uk` redirect to `localhost`? I remember there's some setup involved - don't quite remember what has to be done, or how to search for it. If anyone new comes to the project, they'll likely not have that set up.

Reply:
Aha! Yes i will add this now.